### DISCUSSION TITLE
Why `OutcomeException` is derived from `BaseException` (instead of `Exception`)?

### DISCUSSION THREAD

**Discussion by wangshangsam** [pytest-dev/pytest#12762] [Q&A] [answered]
Hi,

I'm wondering why [OutcomeException](https://github.com/pytest-dev/pytest/blob/4604387ae444a4f3478243f6d12f40b0cc48e121/src/_pytest/outcomes.py#L18) is derived from `BaseException` (instead of `Exception`)? Is it only because that any error derived from `OutcomeException` is not meant to be catched in `try: ... except ...:`, or is there any runtime-related or architectural reason that it has to be `BaseException`?

The reason why I'm asking this question is the following:
- I have been writting some tests to test a piece of Pulumi IaC code.
- In Pulumi, the [language server](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/automation/_server.py#L31) drives a piece of Pulumi code (i.e., the `self.program`), which, in the case of the tests that I'm writing, contains `pytest.raises(...)`.
- The language server [catches all `Exception`s (note: NOT `BaseException` which is a parent class of `Exception`) raised in the `self.program` and reports them](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/automation/_server.py#L97).
- In the event where `pytest.raises(...)` is expecting an error but no error is raised in the code wrapped by `pytest.raises(...)`, [`fail(self.message)`](https://github.com/pytest-dev/pytest/blob/4604387ae444a4f3478243f6d12f40b0cc48e121/src/_pytest/python_api.py#L1015) is called and raises [`Failed`](https://github.com/pytest-dev/pytest/blob/4604387ae444a4f3478243f6d12f40b0cc48e121/src/_pytest/outcomes.py#L178)
- Because `Failed` is a child class of [`OutcomeException`](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/outcomes.py#L64) which is a child class of `BaseException`, Pulumi's language server (which catches `Exception` as mentioned) isn't able to catch this error, which leads to Pulumi's automation runtime thinking `self.program` is fine even though it's not, which then leads to a deadlock.

Thanks!

**Comment by RonnyPfannschmidt** [comment, marked as answer]
Those are indeed derived from base exceptions to prevent accidental shallowing and catching

**Reply by wangshangsam** [reply]
Thanks a lot for the quick answer!

**Comment by The-Compiler** [comment]
See #580 (where this change was originally made) and #12235

**Reply by wangshangsam** [reply]
Really appreciate the references!
